### PR TITLE
copier: Add bind function to configure source/sink buffers params

### PR DESCRIPTION
**Pull request by softwarecki**
Setting the source/sink buffers parameters in the copier_update_params function is not sufficient. If a sink buffer is attached during copier operation, the module will not set buffers parameters. Move the buffer parameter configurations to the new copier_bind bind function. Delete the sink buffers configurations from the copier_module_copy function. There is no need to configure sink buffers parameters on each copy. We are assured that they were configured at the time of bind.

Fixes: #9123

### CODE REVIEW

**Review by iganakov — COMMENTED** [review]
Looks like duplicate of
https://github.com/thesofproject/sof/pull/9197 copier: update buffer format on bind

**Review by kv2019i — APPROVED** [review]

File: src/audio/copier/copier.c [hunk 1]
```diff
@@ -990,6 +987,49 @@ static int copier_get_hw_params(struct comp_dev *dev, struct sof_ipc_stream_para
...
+	struct list_item *list;
+
+	src_id = IPC4_COMP_ID(bu->primary.r.module_id, bu->primary.r.instance_id);
+	sink_id = IPC4_COMP_ID(bu->extension.r.dst_module_id, bu->extension.r.dst_instance_id);
+
+	if (dev->ipc_config.id == src_id) {
+		/* update sink format */
+		list_for_item(list, &dev->bsink_list) {
+			buffer = container_of(list, struct comp_buffer, source_list);
+			uint32_t id = IPC4_SINK_QUEUE_ID(buf_get_id(buffer));
```
Comment (iganakov) [inline src/audio/copier/copier.c:1006]: There is a wrong macro used to retrieve sink id. Buffer id is created by `IPC4_COMP_ID(src_queue, dst_queue)`, in such a case `IPC4_SINK_QUEUE_ID` will return `dst_queue` id and not the `src_queue` id which we are interested in. Thus, you won't be able to update corresponding copier sink for sink num > 1.
`IPC4_SRC_QUEUE_ID()` should be used instead

**Review by iganakov — CHANGES_REQUESTED** [review]

File: src/audio/copier/copier.c [hunk 2]
```diff
@@ -990,6 +987,49 @@ static int copier_get_hw_params(struct comp_dev *dev, struct sof_ipc_stream_para
...
+		}
+	} else if (dev->ipc_config.id == sink_id) {
+		/*
+		 * force update the source buffer format to cover cases where the source module
+		 * fails to set the sink buffer params
+		 */
+		if (!list_is_empty(&dev->bsource_list)) {
+			const struct ipc4_audio_format *const in_fmt = &cd->config.base.audio_fmt;
+
+			buffer = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
```
Comment (lyakh) [inline src/audio/copier/copier.c:1021]: copiers cannot have multiple sources? But yes, it's a copy of the original code

**Review by tmleman — APPROVED** [review]

File: src/audio/copier/copier.c [hunk 3]
```diff
@@ -990,6 +987,49 @@ static int copier_get_hw_params(struct comp_dev *dev, struct sof_ipc_stream_para
...
+static int copier_bind(struct processing_module *mod, void *data)
+{
+	const struct ipc4_module_bind_unbind *const bu = (struct ipc4_module_bind_unbind *)data;
+	struct copier_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *buffer;
+	uint32_t src_id, sink_id;
+	struct list_item *list;
+
+	src_id = IPC4_COMP_ID(bu->primary.r.module_id, bu->primary.r.instance_id);
```
Comment (iganakov) [inline src/audio/copier/copier.c:999]: I think the whole function could be simplified. 
```suggestion
	uint32_t src_queue_id = bu->extension.r.src_queue;
	uint32_t dst_id = IPC4_COMP_ID(bu->extension.r.dst_module_id,
				       bu->extension.r.dst_instance_id);
	if (dev->ipc_config.id == dst_id)
		return 0;

	/* Update corresponding sink buffer format */
	list_for_item(sink_list, &dev->bsink_list) {
		int sink_id;

		sink = container_of(sink_list, struct comp_buffer, source_list);
		sink_id = IPC4_SRC_QUEUE_ID(buf_get_id(sink));
		if (sink_id == src_queue_id) {
			ipc4_update_buffer_format(sink, &cd->out_fmt[sink_id]);
			return 0;
		}
	}

	comp_err(dev, "No sink buffer found for src_queue=%d", src_queue_id);
	return -ENODEV;
```

**Review by iganakov — COMMENTED** [review]
Driver sends SET_SINK_FORMAT IPC to configure copier module sink processing parameters before bind when there is no processing buffer between connected components. Please also remove sink buffer's update from `copier_set_sink_fmt()`. 
It would be nice to simplify copier bind operation, please find my suggestions below

**Review by iganakov — APPROVED** [review]
Looks good to me

**Review by dnikodem — APPROVED** [review]

**Review by gbernatxintel — APPROVED** [review]

**Comment by lgirdwood** [comment]
@softwarecki can you confirm if internal CI is good to merge ? Seeing some red on Jenkins, lets rerun to be sure.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by softwarecki** [comment]
I restored the code that sets the sink parameters in the `copier_update_params` function. Parameters set in the `bind` function can be overwritten by the `comp_verify_params` function called by `ipc4_pipeline_prepare` -> `ipc4_pipeline_params`.

I look forward to the new CI results...

**Comment by lgirdwood** [comment]
> I restored the code that sets the sink parameters in the `copier_update_params` function. Parameters set in the `bind` function can be overwritten by the `comp_verify_params` function called by `ipc4_pipeline_prepare` -> `ipc4_pipeline_params`.
> 
> I look forward to the new CI results...

All look good except for LNL SDW, can you check ? I would expect copier changes to fail on all test (not just SDW), we can rerun again if you think its DUT or testing infra related.

**Comment by kv2019i** [comment]
@lgirdwood wrote:
> All look good except for LNL SDW, can you check ? I would expect copier changes to fail on all test (not just SDW), we can rerun again if you think its DUT or testing infra related.

The LNL SDW is caused by this: https://github.com/thesofproject/linux/issues/5098

**Review by marcinszkudlinski — APPROVED** [review]

**Comment by softwarecki** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
The pause/resume failure in https://sof-ci.01.org/sofpr/PR9276/build6454/devicetest/index.html?model=LNLM_RVP_HDA&testcase=check-pause-resume-capture-100 is addressed by https://github.com/thesofproject/sof-test/pull/1218, please review.

**Comment by softwarecki** [comment]
SOFCI TEST

**Pull request merged**